From d6c7e29f89de2a1f298409829ae7d4b43cb33e96 Mon Sep 17 00:00:00 2001
From: Pavel Grafov <pgrafov@google.com>
Date: Wed, 29 Dec 2021 15:09:38 +0000
Subject: [PATCH] Make ManagedProvisioning hide non-system overlays

This is to prevent tapjacking where an app obscures the
provisioning activity and tricks the user into inflating a work
profile without consent.

Bug: 209611539
Test: manual
Change-Id: I4bda57bd80f34402e0a63554e719dceda60a296f
Merged-In: I4bda57bd80f34402e0a63554e719dceda60a296f
(cherry picked from commit 4b04bf81f88385d9a364a6a6ef88e0025fd0f84a)
Merged-In:I4bda57bd80f34402e0a63554e719dceda60a296f
---
 AndroidManifest.xml                                           | 1 +
 .../managedprovisioning/common/SetupLayoutActivity.java       | 4 ++++
 2 files changed, 5 insertions(+)

diff --git a/AndroidManifest.xml b/AndroidManifest.xml
index 2462d5b0..cb39b714 100644
--- a/AndroidManifest.xml
+++ b/AndroidManifest.xml
@@ -35,6 +35,7 @@
     <uses-permission android:name="android.permission.MANAGE_DEVICE_ADMINS" />
     <uses-permission android:name="android.permission.FOREGROUND_SERVICE"/>
     <uses-permission android:name="android.permission.MANAGE_APP_OPS_MODES"/>
+    <uses-permission android:name="android.permission.HIDE_NON_SYSTEM_OVERLAY_WINDOWS"/>
 
     <uses-permission android:name="android.permission.CHANGE_COMPONENT_ENABLED_STATE" />
     <uses-permission android:name="android.permission.CHANGE_CONFIGURATION" />
diff --git a/src/com/android/managedprovisioning/common/SetupLayoutActivity.java b/src/com/android/managedprovisioning/common/SetupLayoutActivity.java
index 6ae59535..7e63774f 100644
--- a/src/com/android/managedprovisioning/common/SetupLayoutActivity.java
+++ b/src/com/android/managedprovisioning/common/SetupLayoutActivity.java
@@ -17,6 +17,7 @@
 package com.android.managedprovisioning.common;
 
 import static android.view.View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR;
+import static android.view.WindowManager.LayoutParams.SYSTEM_FLAG_HIDE_NON_SYSTEM_OVERLAY_WINDOWS;
 
 import static com.android.internal.logging.nano.MetricsProto.MetricsEvent.VIEW_UNKNOWN;
 
@@ -57,6 +58,9 @@ public abstract class SetupLayoutActivity extends Activity {
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
+
+        getWindow().addSystemFlags(SYSTEM_FLAG_HIDE_NON_SYSTEM_OVERLAY_WINDOWS);
+
         mTimeLogger = new TimeLogger(this, getMetricsCategory());
         mTimeLogger.start();
 
-- 
2.35.1.574.g5d30c73bfb-goog

